sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/highlighting/XMLHighlighting.java

```diff
@@ -0,0 +1,224 @@
+      // empty (or autoclosing) element is raised twice as start and end element, so we need to highlight closing "/" which is placed just before ">"
+      addHighlighting(closingBracketStartOffset - 1, closingBracketStartOffset, "k");
+
+    } else {
+      addHighlighting(startOffset, closingBracketStartOffset + 1, "k");
+    }
+  }
+
+  private void highlightStartElement(XMLStreamReader xmlReader, int startOffset) {
+    int closingBracketStartOffset = getTagClosingBracketStartOffset(startOffset);
+    int endOffset = startOffset + getNameWithNamespaceLength(xmlReader) + 1;
+
+    addHighlighting(startOffset, endOffset, "k");
+    highlightAttributes(endOffset, closingBracketStartOffset);
+    addHighlighting(closingBracketStartOffset, closingBracketStartOffset + 1, "k");
+  }
+
+  private void highlightXmlDeclaration() {
+    if (content.startsWith(XML_DECLARATION_TAG)) {
+      int startOffset = 0;
+      int closingBracketStartOffset = getTagClosingBracketStartOffset(0);
+
+      addHighlighting(startOffset, startOffset + XML_DECLARATION_TAG.length(), "k");
+      highlightAttributes(XML_DECLARATION_TAG.length(), closingBracketStartOffset);
+      addHighlighting(closingBracketStartOffset - 1, closingBracketStartOffset + 1, "k");
+    }
+  }
+
+  private void highlightAttributes(int from, int to) {
+    int counter = from + 1;
+
+    Integer startOffset = null;
+    Character attributeValueQuote = null;
+
+    while (counter < to) {
+      char c = content.charAt(counter);
+
+      if (startOffset == null && !Character.isWhitespace(c)) {
+        startOffset = counter;
+      }
+
+
+      if (attributeValueQuote != null && attributeValueQuote == c) {
+        addHighlighting(startOffset, counter + 1, "s");
+        counter++;
+        startOffset = null;
+        attributeValueQuote = null;
+      }
+
+      if (c == '=' && attributeValueQuote == null) {
+        addHighlighting(startOffset, counter, "c");
+
+        do {
+          counter++;
+          c = content.charAt(counter);
+        } while (c != '\'' && c != '"');
+
+        startOffset = counter;
+        attributeValueQuote = c;
+      }
+
+
+      counter++;
+    }
+  }
+
+  private int getTagClosingBracketStartOffset(int startOffset) {
+    return getClosingBracketStartOffset(startOffset, false);
+  }
+
+  private int getCDATAClosingBracketStartOffset(int startOffset) {
+    return getClosingBracketStartOffset(startOffset, true);
+  }
+
+  private int getClosingBracketStartOffset(int startOffset, boolean isCDATA) {
+    int counter = startOffset + 1;
+    while (startOffset < content.length()) {
+      if (content.charAt(counter) == '>' && bracketsBefore(isCDATA, counter)) {
+        return counter;
+      }
+      counter++;
+    }
+
+    throw new IllegalStateException("No \">\" found.");
+  }
+
+  private boolean bracketsBefore(boolean isCDATA, int counter) {
+    return !isCDATA || content.charAt(counter - 1) == ']' && content.charAt(counter - 2) == ']';
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS864)